near-sdk/src/types/account_id.rs

```diff
@@ -0,0 +1,174 @@
+    Debug,
+    Clone,
+    PartialEq,
+    PartialOrd,
+    Ord,
+    Eq,
+    BorshDeserialize,
+    BorshSerialize,
+    Serialize,
+    Hash,
+    BorshSchema,
+)]
+pub struct AccountId(String);
+
+impl AccountId {
+    /// Returns reference to the account ID bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+    /// Returns reference to the account ID string.
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+    /// Consumes `self` to return inner [`String`] representation.
+    pub fn into_string(self) -> String {
+        self.0
+    }
+    // TODO this should probably be marked as unstable or crate scoped
+    /// Constructs new AccountId from `String` without checking validity.
+    pub fn new_unchecked(id: String) -> Self {
+        Self(id)
+    }
+}
+
+impl fmt::Display for AccountId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<AccountId> for String {
+    fn from(id: AccountId) -> Self {
+        id.0
+    }
+}
+
+impl AsRef<str> for AccountId {
+    fn as_ref(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+impl<'de> Deserialize<'de> for AccountId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, <D as de::Deserializer<'de>>::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s: String = Deserialize::deserialize(deserializer)?;
+        Self::try_from(s).map_err(de::Error::custom)
+    }
+}
+
+impl TryFrom<&str> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        value.parse()
+    }
+}
+
+impl TryFrom<&[u8]> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        core::str::from_utf8(value)
+            .map_err(|_| ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidUtf8 })?
+            .parse()
+    }
+}
+
+fn validate_account_id(id: &str) -> Result<(), ParseAccountIdError> {
+    if is_valid_account_id(id.as_bytes()) {
+        Ok(())
+    } else {
+        Err(ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidAccountId })
+    }
+}
+
+impl TryFrom<Vec<u8>> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_from(
+            String::from_utf8(value)
+                .map_err(|_| ParseAccountIdError { kind: ParseAccountIdErrorKind::InvalidUtf8 })?,
+        )
+    }
+}
+
+impl TryFrom<String> for AccountId {
+    type Error = ParseAccountIdError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        validate_account_id(value.as_str())?;
+        Ok(Self(value))
+    }
+}
+
+impl std::str::FromStr for AccountId {
+    type Err = ParseAccountIdError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        validate_account_id(value)?;
+        Ok(Self(value.to_string()))
+    }
+}
```

Comment:
I think these two are the only two conversions we need. I think it's better to remove the rest of the conversions, as they can already be expressed by the user.

Reply:
yeah I was just trying to reduce the amount of breakages, since the types used previously have these conversions explicitly and would require changes on their end

Reply:
Not sure how I feel about this. The problem is that, if we don't take away this conversions now, we won't be able to remove them in the future. OTOH, it's not they are actively harmful. 

I guess, I'd comment them out, see what the breakage looks in practice, and then decide, biasing to remove. 

No strong opinion here though!

Reply:
I don't see which would be preferred to be removed. I can see the `TryFrom<&str>` but I see a benefit to keeping both of these you linked to in the comment because `FromStr` allows for the easiest conversion and `TryFrom<String>` allows keeping the allocated `String`. Is your point that we should force the re-allocation just because the internal type might be switched from String in the future?

Reply:
I mean the two linked conversions should stay, the *others* (tryFrom &str, &[u8], Vec<u8>) could go.

Reply:
hmm, I suppose that's the case, I just see the benefit of coming from bytes because it's the internal representation when accessing from the runtime. Yeah we could remove it, but if someone wanted to do low-level things, it's just a much harder API to use

I'm fine with removing `&str` but I'd like to keep the bytes ones if possible to make it easier to work directly with registers

Reply:
> I just see the benefit of coming from bytes because it's the internal representation when accessing from the runtime

Note that this is just a limitation of wasm/near API: there are `str` on the other side, we just can't express this. 

But yeah, you point that if someone builds directly against `sys::current_account_id` and friends, they'll use the `TryFrom<Vec<u8>>` API is a good one. 

I guess, overall my personal preference would still be to let the caller to utf8 validation in this case: if you are using low-level things anyway, you probably are not doing this for ergonomics, but for control. And you might want to do things like `Account::new_unchecked(String::from_utf8_unchecked)`.

But not strong opinion here. Having to carry `Utf8Error` is unfortunate, but it's not a big deal

Reply:
I see the point, let me try to trim this down more, I suppose the breakages will be worth this in the long term